README.md
# createlwc README


## Features

less code,more time to deal with affairs.

★Only one command you need to remember: createlwc(or even lwc is fine)

1.you can create lwc with customize XML file.(check where you want to display on)

for example:

![](https://i.imgur.com/OUkLi.gif)

## Requirements

 ・salesforce CLI

## Known Issues

If you find any bugs, please contact me : youkisaki@qq.com

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Added features X, Y, and Z.

---

## Following extension guidelines

Ensure that you've read through the extensions guidelines and follow the best practices for creating your extension.

* [Extension Guidelines](https://code.visualstudio.com/api/references/extension-guidelines)

## Working with Markdown

You can author your README using Visual Studio Code. Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux).
* Toggle preview (`Shift+Cmd+V` on macOS or `Shift+Ctrl+V` on Windows and Linux).
* Press `Ctrl+Space` (Windows, Linux, macOS) to see a list of Markdown snippets.

## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "createlwc" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

out/LMS.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LMSChannel = exports.PUBLMS = exports.SUBLMS = void 0;
class SUBLMS {
    makeSubHtml() {
        const Htmlcode = '<template>' + '\n' +
            '<lightning-card>' + '\n' +
            '\t<h2>here is the passed value</h2>' + '\n' +
            '\t<p class="slds-p-horizontal_small">' + '\n' +
            '\t\t<lightning-input type="text" value={acceptedValue} read-only="true"></lightning-input>' + '\n' +
            '\t</p>' + '\n' +
            '</lightning-card>' + '\n' +
            '</template>';
        return Htmlcode;
    }
    makeSubJS(lwcname, channelname, fieldName) {
        const JScode = 'import { api,wire,LightningElement }from \'lwc\';' + '\n' +
            '//import apexMethodName from \'@salesforce/apex/Classname.apexMethod\';' + '\n' +
            'import { subscribe, MessageContext } from \'lightning/messageService\';' + '\n' +
            'import ' + channelname + 'Channel from \'@salesforce/messageChannel/' + channelname + '__c\';' + '\n' +
            '\n' +
            '\n' +
            'export default class ' + lwcname + ' extends LightningElement {' + '\n' +
            '\n' +
            '\t@api acceptedValue;' + '\n' +
            '\t@wire(MessageContext)' + '\n' +
            '\tmessageContext;' + '\n' +
            '\t//initialization(初期化):Called when the element is inserted into a document' + '\n' +
            '\tconnectedCallback() {' + '\n' +
            '\t\tthis.handleSubscribe();' + '\n' +
            '\t}' +
            '\n' +
            '\t//For overwrite(UIがloading済なので、ここで上書き):Called after every render of the component' + '\n' +
            '\trenderedCallback() {' + '\n' +
            '\n' +
            '\t}' +
            '\n' +
            '\thandleSubscribe() {' + '\n' +
            '    this.designDocSubscription = subscribe(this.messageContext, ' + channelname + 'Channel, (message) => {' + '\n' +
            '      if (message.' + fieldName + ' != undefined) {' + '\n' +
            '       this.acceptedValue = message.' + fieldName + ';' + '\n' +
            '      }' + '\n' +
            '      if (message.forceRefresh && this.cacheData) {' + '\n' +
            '        refreshApex(this.cacheData);' + '\n' +
            '      }' + '\n' +
            '    });' + '\n' +
            '  }' + '\n' +
            '\n' +
            '}';
        return JScode;
    }
    makeXMLcode(lwcname, xml_Pages, objectNM) {
        let XMLcode = '<?xml version="1.0" encoding="UTF-8" ?>' + '\n' +
            '<LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">' + '\n' +
            '\t<apiVersion>55.0</apiVersion>' + '\n' +
            '\t<isExposed>true</isExposed>' + '\n' +
            '\t<masterLabel>' + lwcname + '</masterLabel>' + '\n';
        if (xml_Pages.length !== 0) {
            XMLcode = XMLcode + '\t<targets>' + '\n';
            if (xml_Pages.includes('RecordPage')) {
                XMLcode = XMLcode + '\t\t<target>lightning__RecordPage</target>' + '\n';
            }
            if (xml_Pages.includes('AppPage')) {
                XMLcode = XMLcode + '\t\t<target>lightning__AppPage</target>' + '\n';
            }
            if (xml_Pages.includes('HomePage')) {
                XMLcode = XMLcode + '\t\t<target>lightning__HomePage</target>' + '\n';
            }
            if (xml_Pages.includes('Community')) {
                XMLcode = XMLcode + '\t\t<target>lightningCommunity__Page</target>' + '\n';
            }
            XMLcode = XMLcode + '\t</targets>' + '\n';
        }
        if (objectNM !== null && objectNM !== '') {
            XMLcode = XMLcode + '\t<targetConfigs>' + '\n';
            XMLcode = XMLcode + '\t<targetConfig targets="lightning__RecordPage">' + '\n';
            XMLcode = XMLcode + '\t    <objects>' + '\n';
            XMLcode = XMLcode + '\t        <object>' + objectNM + '</object>' + '\n';
            XMLcode = XMLcode + '\t    </objects>' + '\n';
            XMLcode = XMLcode + '\t</targetConfig>' + '\n';
            XMLcode = XMLcode + '\t/targetConfigs>' + '\n';
        }
        XMLcode = XMLcode + '</LightningComponentBundle>';
        return XMLcode;
    }
}
exports.SUBLMS = SUBLMS;
class PUBLMS {
    makePubHtml() {
        const Htmlcode = '<template>' + '\n' +
            '\t<lightning-card>' + '\n' +
            '\t<h2>press enter key to execute</h2>' + '\n' +
            '\t<p class="slds-p-horizontal_small">' + '\n' +
            '\t\t<lightning-input type="text" onkeyup ={handlePublish} value={sendValue} label="Input some text pass to other lwc by lms"></lightning-input>' + '\n' +
            '\t</p>' + '\n' +
            '</lightning-card>' + '\n' +
            '</template>';
        return Htmlcode;
    }
    makePubJS(lwcname, channelname, fieldName) {
        const JScode = 'import { api,wire,LightningElement }from \'lwc\';' + '\n' +
            '//import apexMethodName from \'@salesforce/apex/Classname.apexMethod\';' + '\n' +
            'import { publish, MessageContext } from \'lightning/messageService\';' + '\n' +
            'import ' + channelname + 'Channel from \'@salesforce/messageChannel/' + channelname + '__c\';' + '\n' +
            '\n' +
            '\n' +
            'export default class ' + lwcname + ' extends LightningElement {' + '\n' +
            '\n' +
            '\t@api sendValue;' + '\n' +
            '\t@wire(MessageContext)' + '\n' +
            '\tmessageContext;' + '\n' +
            '\t//initialization(初期化):Called when the element is inserted into a document' + '\n' +
            '\tconnectedCallback() {' + '\n' +
            '\t\t//publish(this.messageContext, ' + channelname + 'Channel, { ' + fieldName + ': 12}); //you can also do publish in here' + '\n' +
            '\t}' +
            '\n' +
            '\t//For overwrite(UIがloading済なので、ここで上書き):Called after every render of the component' + '\n' +
            '\trenderedCallback() {' + '\n' +
            '\n' +
            '\t}' +
            '\n' +
            '\thandlePublish(event) {' + '\n' +
            '\t    if (event.keyCode === 13) { ' + '\n' +
            '\t      publish(this.messageContext, ' + channelname + 'Channel, { ' + fieldName + ': event.target.value});' + '\n' +
            '\t     };' + '\n' +
            '\t  }' + '\n' +
            '\n' +
            '}';
        return JScode;
    }
    makeXMLcode(lwcname, xml_Pages, objectNM) {
        let XMLcode = '<?xml version="1.0" encoding="UTF-8" ?>' + '\n' +
            '<LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">' + '\n' +
            '\t<apiVersion>55.0</apiVersion>' + '\n' +
            '\t<isExposed>true</isExposed>' + '\n' +
            '\t<masterLabel>' + lwcname + '</masterLabel>' + '\n';
        if (xml_Pages.length !== 0) {
            XMLcode = XMLcode + '\t<targets>' + '\n';
            if (xml_Pages.includes('RecordPage')) {
                XMLcode = XMLcode + '\t\t<target>lightning__RecordPage</target>' + '\n';
            }
            if (xml_Pages.includes('AppPage')) {
                XMLcode = XMLcode + '\t\t<target>lightning__AppPage</target>' + '\n';
            }
            if (xml_Pages.includes('HomePage')) {
                XMLcode = XMLcode + '\t\t<target>lightning__HomePage</target>' + '\n';
            }
            if (xml_Pages.includes('Community')) {
                XMLcode = XMLcode + '\t\t<target>lightningCommunity__Page</target>' + '\n';
            }
            XMLcode = XMLcode + '\t</targets>' + '\n';
        }
        if (objectNM !== null && objectNM !== '') {
            XMLcode = XMLcode + '\t<targetConfigs>' + '\n';
            XMLcode = XMLcode + '\t<targetConfig targets="lightning__RecordPage">' + '\n';
            XMLcode = XMLcode + '\t    <objects>' + '\n';
            XMLcode = XMLcode + '\t        <object>' + objectNM + '</object>' + '\n';
            XMLcode = XMLcode + '\t    </objects>' + '\n';
            XMLcode = XMLcode + '\t</targetConfig>' + '\n';
            XMLcode = XMLcode + '\t/targetConfigs>' + '\n';
        }
        XMLcode = XMLcode + '</LightningComponentBundle>';
        return XMLcode;
    }
}
exports.PUBLMS = PUBLMS;
class LMSChannel {
    makeChannel(fieldName) {
        const Channelcode = '<?xml version="1.0" encoding="UTF-8"?>' + '\n' +
            '<LightningMessageChannel xmlns="http://soap.sforce.com/2006/04/metadata">' + '\n' +
            '\t<description>pass value from one lwc to another</description>' + '\n' +
            '\t<isExposed>true</isExposed>' + '\n' +
            '\t<lightningMessageFields>' + '\n' +
            '\t\t<description>descript what do you want to use for </description>' + '\n' +
            '\t\t<fieldName>' + fieldName + '</fieldName>' + '\n' +
            '\t</lightningMessageFields>' + '\n' +
            '\t<masterLabel>test Message Channel</masterLabel>' + '\n' +
            '</LightningMessageChannel>';
        return Channelcode;
    }
}
exports.LMSChannel = LMSChannel;
//# sourceMappingURL=LMS.js.map

out/NormalLwc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NormalLWC = void 0;
class NormalLWC {
    makeHtmlcode() {
        const Htmlcode = '<template>' + '\n' +
            '<lightning-card>' + '\n' +
            '\n' +
            '</lightning-card>' + '\n' +
            '</template>';
        return Htmlcode;
    }
    makeJScode(lwcname) {
        const JScode = 'import { LightningElement, api } from \'lwc\';' + '\n' +
            '//import apexMethodName from \'@salesforce/apex/Classname.apexMethod\';' + '\n' +
            'import { ShowToastEvent } from \'lightning/platformShowToastEvent\';' + '\n' +
            '\n' +
            '\n' +
            'export default class ' + lwcname + ' extends LightningElement {' + '\n' +
            '\n' +
            '\n' +
            '\t//initialization(初期化):Called when the element is inserted into a document' + '\n' +
            '\tconnectedCallback() {' + '\n' +
            '\n' +
            '\t}' +
            '\n' +
            '\t//For overwrite(UIがloading済なので、ここで上書き):Called after every render of the component' + '\n' +
            '\trenderedCallback() {' + '\n' +
            '\n' +
            '\t}' +
            '\n' +
            '\t//Show error message or success message etc' + '\n' +
            '\t//variant : Valid values are: info (default), success, warning, and error' + '\n' +
            '\t//example:this.showMessage(\'create\',\'you create a record\',\'success\');' + '\n' +
            '\tshowMessage(stitle,smessage,svariant) {' + '\n' +
            '\t\tthis.dispatchEvent(' + '\n' +
            '\t\t\tnew ShowToastEvent({' + '\n' +
            '\t\t\t\ttitle: stitle,' + '\n' +
            '\t\t\t\tmessage: smessage,' + '\n' +
            '\t\t\t\tvariant: svariant' + '\n' +
            '\t\t\t})' + '\n' +
            '\t\t);' + '\n' +
            '\t}' +
            '\n' +
            '}';
        return JScode;
    }
    makeXMLcode(lwcname, xml_Pages, objectNM) {
        let XMLcode = '<?xml version="1.0" encoding="UTF-8" ?>' + '\n' +
            '<LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">' + '\n' +
            '\t<apiVersion>55.0</apiVersion>' + '\n' +
            '\t<isExposed>true</isExposed>' + '\n' +
            '\t<masterLabel>' + lwcname + '</masterLabel>' + '\n';
        if (xml_Pages.length !== 0) {
            XMLcode = XMLcode + '\t<targets>' + '\n';
            if (xml_Pages.includes('RecordPage')) {
                XMLcode = XMLcode + '\t\t<target>lightning__RecordPage</target>' + '\n';
            }
            if (xml_Pages.includes('AppPage')) {
                XMLcode = XMLcode + '\t\t<target>lightning__AppPage</target>' + '\n';
            }
            if (xml_Pages.includes('HomePage')) {
                XMLcode = XMLcode + '\t\t<target>lightning__HomePage</target>' + '\n';
            }
            if (xml_Pages.includes('Community')) {
                XMLcode = XMLcode + '\t\t<target>lightningCommunity__Page</target>' + '\n';
            }
            XMLcode = XMLcode + '\t</targets>' + '\n';
        }
        if (objectNM !== null && objectNM !== '') {
            XMLcode = XMLcode + '\t<targetConfigs>' + '\n';
            XMLcode = XMLcode + '\t<targetConfig targets="lightning__RecordPage">' + '\n';
            XMLcode = XMLcode + '\t    <objects>' + '\n';
            XMLcode = XMLcode + '\t        <object>' + objectNM + '</object>' + '\n';
            XMLcode = XMLcode + '\t    </objects>' + '\n';
            XMLcode = XMLcode + '\t</targetConfig>' + '\n';
            XMLcode = XMLcode + '\t</targetConfigs>' + '\n';
        }
        XMLcode = XMLcode + '</LightningComponentBundle>';
        return XMLcode;
    }
}
exports.NormalLWC = NormalLWC;
//# sourceMappingURL=NormalLwc.js.map

out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require("vscode");
const NormalLwc_1 = require("./NormalLwc");
const LMS_1 = require("./LMS");
const ParentWithChild_1 = require("./ParentWithChild");
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {
    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    console.log('Congratulations, your extension "createlwc" is now active!');
    let disposable = vscode.commands.registerCommand('createlwc.createlwc', () => {
        // The code you place here will be executed every time your command is executed
        // Display a message box to the user
        vscode.window.showInformationMessage('Create lwc from Quick LWC!');
        singleSelectionFunc();
    });
    context.subscriptions.push(disposable);
}
exports.activate = activate;
async function singleSelectionFunc() {
    // singleSelection
    const item = await vscode.window.showQuickPick(['NormalLWC', 'ParentWithChild', 'LMS'], { placeHolder: 'Please Select an item', canPickMany: false });
    if (item === 'NormalLWC') {
        createOne('lwc', []);
    }
    else if (item === 'LMS') {
        let result = await createOne('MessageChannel', []);
        if (result !== undefined) {
            await createOne('subscriber', result);
            await createOne('publisher', result);
        }
    }
    else if (item === 'ParentWithChild') {
        let result = await createOne('ChildLWC', []);
        if (result !== undefined) {
            await createOne('ParentLWC', result);
        }
    }
}
async function createOne(newname, result) {
    let options = {
        prompt: "Label: ",
        placeHolder: "Please input the name of " + newname
    };
    const itemname = await vscode.window.showInputBox(options);
    if (!itemname)
        return;
    let objectNM;
    let xml_Pages;
    if (newname !== 'MessageChannel') {
        //multipleSelections
        xml_Pages = await vscode.window.showQuickPick(['RecordPage', 'AppPage', 'HomePage', 'Community'], { placeHolder: 'Please Select Which scenerio to display', canPickMany: true });
        if (!xml_Pages)
            return;
        if (xml_Pages.length !== 0 && xml_Pages.includes('RecordPage')) {
            let options = {
                prompt: "Label: ",
                placeHolder: "Please input one object name for RecordPage"
            };
            objectNM = await vscode.window.showInputBox(options);
        }
        //create a folder
        if (vscode.workspace.workspaceFolders !== undefined) {
            const wsPath = vscode.workspace.workspaceFolders[0].uri.path + '/force-app/main/default/lwc'; // gets the path of the first workspace folder
            const wsedit = new vscode.WorkspaceEdit();
            const filePath = vscode.Uri.file(wsPath + '/' + itemname + '/' + itemname + '.html');
            const filePath2 = vscode.Uri.file(wsPath + '/' + itemname + '/' + itemname + '.js');
            const filePath3 = vscode.Uri.file(wsPath + '/' + itemname + '/' + itemname + '.js-meta.xml');
            vscode.window.showInformationMessage(filePath.toString());
            wsedit.createFile(filePath, { ignoreIfExists: true });
            wsedit.createFile(filePath2, { ignoreIfExists: true });
            wsedit.createFile(filePath3, { ignoreIfExists: true });
            if (newname === 'lwc') {
                let NormalLWC_temp = new NormalLwc_1.NormalLWC();
                wsedit.insert(filePath, new vscode.Position(0, 0), NormalLWC_temp.makeHtmlcode());
                wsedit.insert(filePath2, new vscode.Position(0, 0), NormalLWC_temp.makeJScode(itemname));
                wsedit.insert(filePath3, new vscode.Position(0, 0), NormalLWC_temp.makeXMLcode(itemname, xml_Pages, objectNM ? objectNM : ''));
            }
            else if (newname === 'subscriber') {
                let LMS_temp = new LMS_1.SUBLMS();
                wsedit.insert(filePath, new vscode.Position(0, 0), LMS_temp.makeSubHtml());
                wsedit.insert(filePath2, new vscode.Position(0, 0), LMS_temp.makeSubJS(itemname, result[1], result[0])); //result[0] = fieldname;result[1]= lastTimeitemname
                wsedit.insert(filePath3, new vscode.Position(0, 0), LMS_temp.makeXMLcode(itemname, xml_Pages, objectNM ? objectNM : ''));
            }
            else if (newname === 'publisher') {
                let LMS_temp = new LMS_1.PUBLMS();
                wsedit.insert(filePath, new vscode.Position(0, 0), LMS_temp.makePubHtml());
                wsedit.insert(filePath2, new vscode.Position(0, 0), LMS_temp.makePubJS(itemname, result[1], result[0])); //result[0] = fieldname;result[1]= lastTimeitemname
                wsedit.insert(filePath3, new vscode.Position(0, 0), LMS_temp.makeXMLcode(itemname, xml_Pages, objectNM ? objectNM : ''));
            }
            else if (newname === 'ChildLWC') {
                let LMS_temp = new ParentWithChild_1.CHILD();
                wsedit.insert(filePath, new vscode.Position(0, 0), LMS_temp.makeCHILDHtml());
                wsedit.insert(filePath2, new vscode.Position(0, 0), LMS_temp.makeCHILDJS(itemname)); //result[0] = fieldname;result[1]= lastTimeitemname
                wsedit.insert(filePath3, new vscode.Position(0, 0), LMS_temp.makeXMLcode(itemname, xml_Pages, objectNM ? objectNM : ''));
            }
            else if (newname === 'ParentLWC') {
                let LMS_temp = new ParentWithChild_1.Parent();
                wsedit.insert(filePath, new vscode.Position(0, 0), LMS_temp.makeParentHtml(result[1]));
                wsedit.insert(filePath2, new vscode.Position(0, 0), LMS_temp.makeParentJS(itemname, result[1])); //result[0] = fieldname;result[1]= lastTimeitemname
                wsedit.insert(filePath3, new vscode.Position(0, 0), LMS_temp.makeXMLcode(itemname, xml_Pages, objectNM ? objectNM : ''));
            }
            await vscode.workspace.applyEdit(wsedit);
            vscode.workspace.saveAll();
            vscode.window.showInformationMessage('Create a normal lwc');
            return ['', itemname];
        }
    }
    else {
        let options = {
            prompt: "Label: ",
            placeHolder: "Please input the name of parameter to pass the value"
        };
        let fieldname = await vscode.window.showInputBox(options);
        if (!fieldname)
            return;
        //create a folder
        if (vscode.workspace.workspaceFolders !== undefined) {
            const wsPath = vscode.workspace.workspaceFolders[0].uri.path; // gets the path of the first workspace folder
            const wsedit = new vscode.WorkspaceEdit();
            const filePath = vscode.Uri.file(wsPath + '/messageChannels/' + itemname + '.messageChannel-meta.xml');
            vscode.window.showInformationMessage(filePath.toString());
            wsedit.createFile(filePath, { ignoreIfExists: true });
            let LMS_temp = new LMS_1.LMSChannel();
            wsedit.insert(filePath, new vscode.Position(0, 0), LMS_temp.makeChannel(fieldname));
            await vscode.workspace.applyEdit(wsedit);
            vscode.workspace.saveAll();
            vscode.window.showInformationMessage('Create a channel');
        }
        return [fieldname, itemname];
    }
}
// this method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map

src/LMS.ts
export class SUBLMS {

     makeSubHtml() {
        const Htmlcode ='<template>'+'\n'+
                            '<lightning-card>'+'\n'+
                            '\t<h2>here is the passed value</h2>'+'\n'+
                            '\t<p class="slds-p-horizontal_small">'+'\n'+
                            '\t\t<lightning-input type="text" value={acceptedValue} read-only="true"></lightning-input>'+'\n'+
                            '\t</p>'+'\n'+
                            '</lightning-card>'+'\n'+
                        '</template>';
        return Htmlcode;
    }
 
    makeSubJS(lwcname : string,channelname : string,fieldName : string) {
        const JScode ='import { api,wire,LightningElement }from \'lwc\';'+'\n'+
                        '//import apexMethodName from \'@salesforce/apex/Classname.apexMethod\';'+'\n'+
                        'import { subscribe, MessageContext } from \'lightning/messageService\';'+'\n'+
                        'import '+channelname+'Channel from \'@salesforce/messageChannel/'+ channelname +'__c\';'+'\n'+
                        '\n'+
                        '\n'+
                        'export default class '+ lwcname +' extends LightningElement {'+'\n'+
                        '\n'+
                        '\t@api acceptedValue;'+'\n'+
                        '\t@wire(MessageContext)'+'\n'+
                        '\tmessageContext;'+'\n'+
                        '\t//initialization(初期化):Called when the element is inserted into a document'+'\n'+
                        '\tconnectedCallback() {'+'\n'+
                        '\t\tthis.handleSubscribe();'+'\n'+
                        '\t}'+
                        '\n'+
                        '\t//For overwrite(UIがloading済なので、ここで上書き):Called after every render of the component'+'\n'+
                        '\trenderedCallback() {'+'\n'+
                        '\n'+
                        '\t}'+
                        '\n'+
                        '\thandleSubscribe() {'+'\n'+
                        '    this.designDocSubscription = subscribe(this.messageContext, '+channelname+'Channel, (message) => {'+'\n'+
                        '      if (message.'+fieldName+' != undefined) {'+'\n'+
                        '       this.acceptedValue = message.'+fieldName+';'+'\n'+
                        '      }'+'\n'+
                        '      if (message.forceRefresh && this.cacheData) {'+'\n'+
                        '        refreshApex(this.cacheData);'+'\n'+
                        '      }'+'\n'+
                        '    });'+'\n'+
                        '  }'+'\n'+
                        '\n'+
                        '}';
        return JScode;
    }

    makeXMLcode(lwcname : string, xml_Pages: string[],objectNM:string) {
        let XMLcode ='<?xml version="1.0" encoding="UTF-8" ?>'+'\n'+
                        '<LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">'+'\n'+
                        '\t<apiVersion>55.0</apiVersion>'+'\n'+
                        '\t<isExposed>true</isExposed>'+'\n'+
                        '\t<masterLabel>' + lwcname + '</masterLabel>'+'\n';
                        if (xml_Pages.length !== 0) {
                            XMLcode = XMLcode +'\t<targets>'+'\n';
                            if(xml_Pages.includes('RecordPage')){
                                XMLcode = XMLcode +  '\t\t<target>lightning__RecordPage</target>'+'\n';
                            }
                            if(xml_Pages.includes('AppPage')){
                                XMLcode = XMLcode +  '\t\t<target>lightning__AppPage</target>'+'\n';
                            }
                            if(xml_Pages.includes('HomePage')){
                                XMLcode = XMLcode +  '\t\t<target>lightning__HomePage</target>'+'\n';
                            }
                            if(xml_Pages.includes('Community')){
                                XMLcode = XMLcode +  '\t\t<target>lightningCommunity__Page</target>'+'\n';
                            }
                            XMLcode = XMLcode +'\t</targets>'+'\n';
                        }
                        if(objectNM !== null && objectNM !== ''){
                            XMLcode = XMLcode +  '\t<targetConfigs>'+'\n';
                            XMLcode = XMLcode +  '\t<targetConfig targets="lightning__RecordPage">'+'\n';
                            XMLcode = XMLcode +  '\t    <objects>'+'\n';
                            XMLcode = XMLcode +  '\t        <object>'+ objectNM +'</object>'+'\n';
                            XMLcode = XMLcode +  '\t    </objects>'+'\n';
                            XMLcode = XMLcode +  '\t</targetConfig>'+'\n';
                            XMLcode = XMLcode +  '\t/targetConfigs>'+'\n';
                        }
            XMLcode = XMLcode +'</LightningComponentBundle>';
        return XMLcode;
    }

}



export class PUBLMS {

    makePubHtml() {
       const Htmlcode ='<template>'+'\n'+
                           '\t<lightning-card>'+'\n'+
                           '\t<h2>press enter key to execute</h2>'+'\n'+
                           '\t<p class="slds-p-horizontal_small">'+'\n'+
                           '\t\t<lightning-input type="text" onkeyup ={handlePublish} value={sendValue} label="Input some text pass to other lwc by lms"></lightning-input>'+'\n'+
                           '\t</p>'+'\n'+
                           '</lightning-card>'+'\n'+
                       '</template>';
       return Htmlcode;
   }

   makePubJS(lwcname : string,channelname : string,fieldName : string) {
       const JScode ='import { api,wire,LightningElement }from \'lwc\';'+'\n'+
                       '//import apexMethodName from \'@salesforce/apex/Classname.apexMethod\';'+'\n'+
                       'import { publish, MessageContext } from \'lightning/messageService\';'+'\n'+
                       'import '+channelname+'Channel from \'@salesforce/messageChannel/'+ channelname +'__c\';'+'\n'+
                       '\n'+
                       '\n'+
                       'export default class '+ lwcname +' extends LightningElement {'+'\n'+
                       '\n'+
                       '\t@api sendValue;'+'\n'+
                       '\t@wire(MessageContext)'+'\n'+
                       '\tmessageContext;'+'\n'+
                       '\t//initialization(初期化):Called when the element is inserted into a document'+'\n'+
                       '\tconnectedCallback() {'+'\n'+
                       '\t\t//publish(this.messageContext, '+channelname+'Channel, { '+fieldName+': 12}); //you can also do publish in here'+'\n'+
                       '\t}'+
                       '\n'+
                       '\t//For overwrite(UIがloading済なので、ここで上書き):Called after every render of the component'+'\n'+
                       '\trenderedCallback() {'+'\n'+
                       '\n'+
                       '\t}'+
                       '\n'+
                       '\thandlePublish(event) {'+'\n'+
                       '\t    if (event.keyCode === 13) { '+'\n'+
                       '\t      publish(this.messageContext, '+channelname+'Channel, { '+fieldName+': event.target.value});'+'\n'+
                       '\t     };'+'\n'+
                       '\t  }'+'\n'+
                       '\n'+
                       '}';
       return JScode;
   }




   makeXMLcode(lwcname : string, xml_Pages: string[],objectNM:string) {
       let XMLcode ='<?xml version="1.0" encoding="UTF-8" ?>'+'\n'+
                       '<LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">'+'\n'+
                       '\t<apiVersion>55.0</apiVersion>'+'\n'+
                       '\t<isExposed>true</isExposed>'+'\n'+
                       '\t<masterLabel>' + lwcname + '</masterLabel>'+'\n';
                       if (xml_Pages.length !== 0) {
                           XMLcode = XMLcode +'\t<targets>'+'\n';
                           if(xml_Pages.includes('RecordPage')){
                               XMLcode = XMLcode +  '\t\t<target>lightning__RecordPage</target>'+'\n';
                           }
                           if(xml_Pages.includes('AppPage')){
                               XMLcode = XMLcode +  '\t\t<target>lightning__AppPage</target>'+'\n';
                           }
                           if(xml_Pages.includes('HomePage')){
                               XMLcode = XMLcode +  '\t\t<target>lightning__HomePage</target>'+'\n';
                           }
                           if(xml_Pages.includes('Community')){
                               XMLcode = XMLcode +  '\t\t<target>lightningCommunity__Page</target>'+'\n';
                           }
                           XMLcode = XMLcode +'\t</targets>'+'\n';
                       }
                       if(objectNM !== null && objectNM !== ''){
                           XMLcode = XMLcode +  '\t<targetConfigs>'+'\n';
                           XMLcode = XMLcode +  '\t<targetConfig targets="lightning__RecordPage">'+'\n';
                           XMLcode = XMLcode +  '\t    <objects>'+'\n';
                           XMLcode = XMLcode +  '\t        <object>'+ objectNM +'</object>'+'\n';
                           XMLcode = XMLcode +  '\t    </objects>'+'\n';
                           XMLcode = XMLcode +  '\t</targetConfig>'+'\n';
                           XMLcode = XMLcode +  '\t/targetConfigs>'+'\n';
                       }
           XMLcode = XMLcode +'</LightningComponentBundle>';
       return XMLcode;
   }

}

export class LMSChannel {
        makeChannel(fieldName : string) {
            const Channelcode ='<?xml version="1.0" encoding="UTF-8"?>'+'\n'+
                                '<LightningMessageChannel xmlns="http://soap.sforce.com/2006/04/metadata">'+'\n'+
                                '\t<description>pass value from one lwc to another</description>'+'\n'+
                                '\t<isExposed>true</isExposed>'+'\n'+
                                '\t<lightningMessageFields>'+'\n'+
                                '\t\t<description>descript what do you want to use for </description>'+'\n'+
                                '\t\t<fieldName>' + fieldName + '</fieldName>'+'\n'+
                                '\t</lightningMessageFields>'+'\n'+
                                '\t<masterLabel>test Message Channel</masterLabel>'+'\n'+
                                '</LightningMessageChannel>';
            return Channelcode;
}
}